File: src/test/java/umc/th/juinjang/api/pencil/service/AcquiredPencilServiceTest.java
package umc.th.juinjang.api.pencil.service;

import static org.assertj.core.api.Assertions.*;

import java.time.LocalDateTime;
import java.util.List;

import org.assertj.core.groups.Tuple;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;

import umc.th.juinjang.api.IntegrationTestSupport;
import umc.th.juinjang.api.pencil.service.response.AcquiredPencilResponse;
import umc.th.juinjang.domain.member.model.Member;
import umc.th.juinjang.domain.member.repository.MemberRepository;
import umc.th.juinjang.domain.pencil.acquired.model.AcquiredPencil;
import umc.th.juinjang.domain.pencil.acquired.model.AcquiredType;
import umc.th.juinjang.domain.pencil.acquired.repository.AcquiredPencilRepository;
import umc.th.juinjang.testutil.fixture.MemberFixture;

class AcquiredPencilServiceTest extends IntegrationTestSupport {

	@Autowired
	private MemberRepository memberRepository;

	@Autowired
	private AcquiredPencilRepository acquiredPencilRepository;

	@Autowired
	private AcquiredPencilService pencilService;

	@AfterEach
	void tearDown() {
		acquiredPencilRepository.deleteAllInBatch();
		memberRepository.deleteAllInBatch();
	}

	@DisplayName("얻은 연필 목록이 없는 경우에는 빈 배열이 반환된다.")
	@Test
	void getEmptyAcquiredPencilsList() {
		// given
		Member member = MemberFixture.createDefaultMember();
		memberRepository.save(member);

		// when
		List<AcquiredPencilResponse> list = pencilService.getAcquiredPencils(member);

		// then
		assertThat(list).hasSize(0);
	}

	@DisplayName("얻은 연필 목록이 생성 시간(createdAt) 내림차순으로 정렬되어 반환된다.")
	@Test
	void getAcquiredPencilsOrderedByCreatedAtDesc() {
		// given
		Member member = MemberFixture.createDefaultMember();
		memberRepository.save(member);

		// 시간을 내림차순으로 생성 (최신 시간이 먼저 오도록)
		LocalDateTime now = LocalDateTime.now();
		LocalDateTime time1 = now.minusHours(4);
		LocalDateTime time2 = now.minusHours(3);
		LocalDateTime time3 = now.minusHours(2);
		LocalDateTime time4 = now.minusHours(1);
		LocalDateTime time5 = now;

		// 명확한 순서로 데이터 생성 (시간 역순으로)
		AcquiredPencil pencil1 = createAcquiredPencilWithTime(time1, "노트 작성으로 연필 획득", 1L, 10L, false, AcquiredType.NOTE,
			member);
		AcquiredPencil pencil2 = createAcquiredPencilWithTime(time2, "연필팩 구매로 연필 추가", 2L, 20L, true, AcquiredType.SOLD,
			member);
		AcquiredPencil pencil3 = createAcquiredPencilWithTime(time3, "매물 판매로 연필 획득", 3L, 30L, false, AcquiredType.SOLD,
			member);
		AcquiredPencil pencil4 = createAcquiredPencilWithTime(time4, "다른 노트 작성으로 연필 획득", 4L, 15L, true,
			AcquiredType.NOTE, member);
		AcquiredPencil pencil5 = createAcquiredPencilWithTime(time5, "또 다른 매물 판매로 연필 획득", 5L, 25L, false,
			AcquiredType.SOLD, member);

		acquiredPencilRepository.saveAll(List.of(pencil1, pencil2, pencil3, pencil4, pencil5));

		// when
		List<AcquiredPencilResponse> foundPencils = pencilService.getAcquiredPencils(member);

		// then
		assertThat(foundPencils).hasSize(5)
			.extracting("content", "sharedNoteId", "acquiredQuantity")
			.containsExactly(
				// 최신 시간부터 나열 (내림차순)
				Tuple.tuple("또 다른 매물 판매로 연필 획득", 5L, 25L),
				Tuple.tuple("다른 노트 작성으로 연필 획득", 4L, 15L),
				Tuple.tuple("매물 판매로 연필 획득", 3L, 30L),
				Tuple.tuple("연필팩 구매로 연필 추가", 2L, 20L),
				Tuple.tuple("노트 작성으로 연필 획득", 1L, 10L)
			);
	}

	private AcquiredPencil createAcquiredPencilWithTime(LocalDateTime createdAt, String content, Long sharedNoteId,
		Long acquiredQuantity, boolean isRead, AcquiredType type, Member member) {
		return AcquiredPencil.createWithDate(member, content, sharedNoteId, acquiredQuantity, isRead, type, createdAt);
	}
}

File: apps/juinjang-api/src/test/kotlin/umc/th/juinjang/application/service/acquired/AcquiredPencilServiceTest.kt
package umc.th.juinjang.application.service.acquired

import jakarta.persistence.EntityNotFoundException
import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.Assertions.assertThatThrownBy
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Nested
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import umc.th.juinjang.domain.pencil.acquired.AcquiredPencil
import umc.th.juinjang.domain.pencil.acquired.AcquiredType
import umc.th.juinjang.infrastructure.pencil.acquired.AcquiredPencilJpaRepository
import umc.th.juinjang.support.IntegrationTestSupport

class AcquiredPencilServiceTest : IntegrationTestSupport() {

    @Autowired
    private lateinit var pencilService: AcquiredPencilService

    @Autowired
    private lateinit var acquiredPencilJpaRepository: AcquiredPencilJpaRepository

    @AfterEach
    fun tearDown() {
        acquiredPencilJpaRepository.deleteAllInBatch()
    }

    @Nested
    @DisplayName("getAcquiredPencilsByMember")
    inner class GetAcquiredPencilsByMember {

        @Test
        @DisplayName("멤버의 연필 히스토리 목록을 반환한다")
        fun returnsList() {
            // arrange
            val memberId = 1L
            acquiredPencilJpaRepository.save(createPencil(memberId = memberId, content = "노트 공유 보상"))
            acquiredPencilJpaRepository.save(createPencil(memberId = memberId, content = "광고 시청 보상"))

            // act
            val result = pencilService.getAcquiredPencilsByMember(memberId)

            // assert
            assertThat(result).hasSize(2)
        }

        @Test
        @DisplayName("히스토리가 없으면 빈 목록을 반환한다")
        fun returnsEmptyList() {
            // act
            val result = pencilService.getAcquiredPencilsByMember(999L)

            // assert
            assertThat(result).isEmpty()
        }
    }

    @Nested
    @DisplayName("findById")
    inner class FindById {

        @Test
        @DisplayName("ID로 연필 히스토리를 조회한다")
        fun returnsEntity() {
            // arrange
            val saved = acquiredPencilJpaRepository.save(
                createPencil(memberId = 1L, content = "노트 공유 보상"),
            )

            // act
            val result = pencilService.findById(saved.id!!)

            // assert
            assertThat(result.id).isEqualTo(saved.id)
            assertThat(result.content).isEqualTo("노트 공유 보상")
        }

        @Test
        @DisplayName("존재하지 않는 ID면 EntityNotFoundException을 던진다")
        fun throwsWhenNotFound() {
            // act & assert
            assertThatThrownBy { pencilService.findById(999L) }
                .isInstanceOf(EntityNotFoundException::class.java)
                .hasMessageContaining("999")
        }
    }

    @Nested
    @DisplayName("existsByMemberIdAndIsReadFalse")
    inner class ExistsByMemberIdAndIsReadFalse {

        @Test
        @DisplayName("안 읽은 히스토리가 있으면 true를 반환한다")
        fun returnsTrue() {
            // arrange
            val memberId = 1L
            acquiredPencilJpaRepository.save(createPencil(memberId = memberId, content = "보상"))

            // act
            val result = pencilService.existsByMemberIdAndIsReadFalse(memberId)

            // assert
            assertThat(result).isTrue()
        }

        @Test
        @DisplayName("모두 읽었으면 false를 반환한다")
        fun returnsFalse() {
            // act (데이터 없음)
            val result = pencilService.existsByMemberIdAndIsReadFalse(1L)

            // assert
            assertThat(result).isFalse()
        }
    }

    @Nested
    @DisplayName("markAsRead")
    inner class MarkAsRead {

        @Test
        @DisplayName("읽음 처리 후 true를 반환하고 DB에 반영된다")
        fun returnsTrueAndMarksRead() {
            // arrange
            val saved = acquiredPencilJpaRepository.save(
                createPencil(memberId = 1L, content = "노트 공유 보상"),
            )

            // act
            val result = pencilService.markAsRead(saved.id!!)

            // assert
            assertThat(result).isTrue()
            val reloaded = acquiredPencilJpaRepository.findById(saved.id!!).orElseThrow()
            assertThat(reloaded.isRead).isTrue()
        }
    }

    @Nested
    @DisplayName("isAllRead")
    inner class IsAllRead {

        @Test
        @DisplayName("모두 읽었으면 true를 반환한다")
        fun returnsTrue() {
            // act (데이터 없음)
            val result = pencilService.isAllRead(1L)

            // assert
            assertThat(result).isTrue()
        }

        @Test
        @DisplayName("안 읽은 것이 있으면 false를 반환한다")
        fun returnsFalse() {
            // arrange
            val memberId = 1L
            acquiredPencilJpaRepository.save(createPencil(memberId = memberId, content = "보상"))

            // act
            val result = pencilService.isAllRead(memberId)

            // assert
            assertThat(result).isFalse()
        }
    }

    private fun createPencil(memberId: Long, content: String): AcquiredPencil =
        AcquiredPencil.create(
            memberId = memberId,
            content = content,
            sharedNoteId = null,
            acquiredQuantity = 5L,
            type = AcquiredType.NOTE,
        )
}
